File: Objective-C/Objective-C/DataUtils.m
//
//  DataUtils.m
//  Objective-C
//
//  Created by Nguyen Anh Minh on 2/19/16.
//  Copyright © 2016 Minh Nguyen. All rights reserved.
//

#import <CoreData/CoreData.h>

#import "DataUtils.h"

@interface DataUtils()

@property (strong, nonatomic) NSManagedObjectContext *managedObjectContext;
@property (strong, nonatomic) NSManagedObjectModel *managedObjectModel;
@property (strong, nonatomic) NSPersistentStoreCoordinator *persistentStoreCoordinator;

@end

@implementation DataUtils

#pragma mark - Init Function

+ (instancetype)getInstance{
    
    static DataUtils *instance;
    static dispatch_once_t token;
    
    dispatch_once(&token, ^{
        instance = [[DataUtils alloc] init];
    });
    
    return instance;
}

+ (void)installDatabaseWithName:(NSString *)name{
    DataUtils *db = [DataUtils getInstance];
    
    if (!db.managedObjectModel) {
        NSURL *modelURL = [[NSBundle mainBundle] URLForResource:name withExtension:@"momd"];
        if (!modelURL) {
            NSLog(@"Couldn't locate model file");
            return;
        }
        db.managedObjectModel = [[NSManagedObjectModel alloc]initWithContentsOfURL:modelURL];
    }
    
    if (!db.persistentStoreCoordinator) {
        db.persistentStoreCoordinator = [[NSPersistentStoreCoordinator alloc]initWithManagedObjectModel:db.managedObjectModel];
        NSURL *storeURL = [[db applicationDocumentsDirectory] URLByAppendingPathComponent:[NSString stringWithFormat:@"%@.sqlite", name]];
        NSError *error = nil;
        [db.persistentStoreCoordinator addPersistentStoreWithType:NSSQLiteStoreType configuration:nil URL:storeURL options:nil error:&error];
        if (error) {
            NSLog(@"%@", error.localizedDescription);
            return;
        }
    }
    
    if (!db.managedObjectContext) {
        db.managedObjectContext = [[NSManagedObjectContext alloc]init];
        [db.managedObjectContext setPersistentStoreCoordinator:db.persistentStoreCoordinator];
    }
}

- (NSURL*)applicationDocumentsDirectory{
    return [[[NSFileManager defaultManager] URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask] lastObject];
}

#pragma mark - Common Function
- (NSArray*)fetchEntityWithName:(NSString *)name
                  withCondition:(NSPredicate *)condition
              withSortCondition:(NSArray*)sortConditions
                      withLimit:(NSInteger)limit{
    
    NSError *error = nil;
    NSFetchRequest *fetchRequest = [[NSFetchRequest alloc]initWithEntityName:name];
    fetchRequest.returnsObjectsAsFaults = NO;
    if (limit > 0) {
        [fetchRequest setFetchLimit:limit];
    }
    
    if (condition) {
        [fetchRequest setPredicate:condition];
    }
    if (sortConditions && sortConditions.count > 0) {
        [fetchRequest setSortDescriptors:sortConditions];
    }
    NSArray *result = [self.managedObjectContext executeFetchRequest:fetchRequest error:&error];
    
    return error || !result ? @[] : result;
}

- (NSArray *)fetchEntityWithName:(NSString *)name withCondition:(NSPredicate *)condition withSortCondition:(NSArray *)sortConditions withLimit:(NSInteger)limit withBindClass:(NSString *)className{
    
    NSArray *objects = [self fetchEntityWithName:name withCondition:condition withSortCondition:sortConditions withLimit:limit];
    if (!className) {
        return objects;
    }
    NSMutableArray *conversion = [NSMutableArray array];
    [objects enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        
        NSArray *keys = [[[obj entity] attributesByName] allKeys];
        NSDictionary *dict = [obj dictionaryWithValuesForKeys:keys];
        
        NSObject *bind = [NSClassFromString(className) new];
        [bind setValuesForKeysWithDictionary:dict];
        
        [conversion addObject:bind];
        
    }];
    
    return conversion;
}

- (BOOL)updateEntity:(NSString*)name
       withCondition:(NSPredicate*)condition
           withValue:(NSDictionary*)values{
    
    NSManagedObject *entity = nil;
    if (condition) {
        NSArray *existingEntities = [self fetchEntityWithName:name withCondition:condition withSortCondition:nil withLimit:-1];
        entity = existingEntities.firstObject;
    }
    if (!entity) {
        entity = [NSEntityDescription insertNewObjectForEntityForName:name inManagedObjectContext:self.managedObjectContext];
    }
    for (NSString* key in values.allKeys) {
        [entity setValue:values[key] forKey:key];
    }
    return [self save];
}

- (BOOL)deleteEntity:(NSString *)name withCondition:(NSPredicate *)condition{
    NSArray *result = [self fetchEntityWithName:name withCondition:condition withSortCondition:nil withLimit:-1];
    for (NSManagedObject *object in result){
        [self.managedObjectContext deleteObject:object];
    }
    return [self save];
}

- (BOOL)save{
    if (self.managedObjectContext) {
        NSError *error = nil;
        if ([self.managedObjectContext  hasChanges]) {
            [self.managedObjectContext save:&error];
        }
        return !error;
    }
    return NO;
}

@end

File: SwiftSource/SwiftSource/DataUtils.swift
//
//  DataUtils.swift
//  SwiftSource
//
//  Created by Nguyen Anh Minh on 2/22/16.
//  Copyright © 2016 Minh Nguyen. All rights reserved.
//

import Foundation
import CoreData

extension NSObject {
    
    class func fromClassName(className : String) -> NSObject {
        let className = NSBundle.mainBundle().infoDictionary!["CFBundleName"] as! String + "." + className
        let aClass = NSClassFromString(className) as! NSObject.Type
        return aClass.init()
    }
    
}

class DataUtils: NSObject {
    
    private var managedObjectContext: NSManagedObjectContext?
    private var managedObjectModel: NSManagedObjectModel?
    private var persistentStoreCoordinator: NSPersistentStoreCoordinator?
    
    // MARK: Setup func
    class var getInstance: DataUtils {
        struct Static {
            static var token = 0
            static var instance: DataUtils? = nil
        }
        
        dispatch_once(&Static.token) { () -> Void in
            Static.instance = DataUtils()
        }
        
        return Static.instance!
    }
    
    class func installDatabase(name: String!) {
        let instance = DataUtils.getInstance
        if (instance.managedObjectModel == nil){
            let modelURL = NSBundle.mainBundle().URLForResource(name, withExtension: "momd")
            if (modelURL == nil){
                return
            }
            instance.managedObjectModel = NSManagedObjectModel(contentsOfURL: modelURL!)
        }
        
        if (instance.persistentStoreCoordinator == nil){
            instance.persistentStoreCoordinator = NSPersistentStoreCoordinator(managedObjectModel: instance.managedObjectModel!)
            var storeURL = instance.applicationDocumentDirectory()
            if (storeURL == nil){
                return
            }
            storeURL = storeURL!.URLByAppendingPathComponent("\(name).sqlite")
            do {
                try instance.persistentStoreCoordinator!.addPersistentStoreWithType(NSSQLiteStoreType, configuration: nil, URL: storeURL, options: nil)
            } catch _ as NSError{
                return
            }
        }
        
        if (instance.managedObjectContext == nil){
            instance.managedObjectContext = NSManagedObjectContext()
            instance.managedObjectContext!.persistentStoreCoordinator = instance.persistentStoreCoordinator
        }
    }
    
    func applicationDocumentDirectory()->NSURL?{
        return NSFileManager.defaultManager().URLsForDirectory(NSSearchPathDirectory.DocumentDirectory, inDomains: NSSearchPathDomainMask.UserDomainMask).last
    }
    
    // MARK: Fetch Entity
    func fetchEntity(name: String!, condition: NSPredicate?, sortCondition: [NSSortDescriptor]?, limit: NSInteger)->[AnyObject]{
        let fetchRequest = NSFetchRequest(entityName: name)
        
        if (limit > 0){
            fetchRequest.fetchLimit = limit
        }
        
        if (condition != nil){
            fetchRequest.predicate = condition!
        }
        
        if (sortCondition != nil){
            fetchRequest.sortDescriptors = sortCondition!
        }
        
        var result = [AnyObject]()
        
        do {
            try result = self.managedObjectContext!.executeFetchRequest(fetchRequest)
        } catch _ as NSError{
            
        }
        return result
    }
    
    func fetchEntity(name: String!, condition: NSPredicate?, sortCondition: [NSSortDescriptor]?, limit: NSInteger, bind: String?)->[AnyObject]{
        var result = self.fetchEntity(name, condition: condition, sortCondition: sortCondition, limit: limit)
        
        if (bind != nil){
            let count = result.count - 1
            if (count == -1){
                return result
            }
            for i in 0...count{
                let object: NSManagedObject = result[i] as! NSManagedObject
                let instance = NSObject.fromClassName(bind!)
                let keys = Array(object.entity.attributesByName.keys)
                let dict = object.dictionaryWithValuesForKeys(keys)
                instance.setValuesForKeysWithDictionary(dict)
                result[i] = instance
            }
        }
        
        return result
    }
    
    // MARK: Update Entity
    func updateEntity(name: String!, condition: NSPredicate?, values: [String:AnyObject])->Bool{
        var entity: NSManagedObject? = nil
        
        if (condition != nil) {
            entity = self.fetchEntity(name, condition: condition, sortCondition: nil, limit: -1).last as? NSManagedObject
        }
        
        if (entity == nil){
            entity = NSEntityDescription.insertNewObjectForEntityForName(name, inManagedObjectContext: self.managedObjectContext!)
        }
        
        entity!.setValuesForKeysWithDictionary(values)
        
        return self.save()
    }
    
    // MARK: Delete Entity
    func deleteEntity(name: String!, condition: NSPredicate?)->Bool{
        let existingObject = self.fetchEntity(name, condition: condition, sortCondition: nil, limit: -1)
        
        for object in existingObject{
            self.managedObjectContext?.deleteObject(object as! NSManagedObject)
        }
        
        return self.save()
    }
    
    // MARK: Save
    func save()->Bool{
        if (self.managedObjectContext != nil){
            do {
                try self.managedObjectContext!.save()
            } catch _ as NSError{
                return false
            }
        }
        return true
    }
    
}
